### PR TITLE
Enhanced route_check.py for multi_asic platforms

### PR DESCRIPTION
**Pull request by deepak-singhal0408**
#### What I did
Enhanced route_check.py script to cover multi-asic platforms.
Accordingly enhanced the test files as well.

MSFT ADO: 25416673

#### How I did it
Enhanced the route_check.py script to take additional optional parameter(--n/namespace). Without this parameter, the check will be run on all asics in multi-asic platforms.
Different connections to DBs are modified accordingly to connect to relevant ns dbs.
Result will be encapsulated under different namespace. For single asic, results will be displayed under Default-Namespace("").

testData and the testfiles are enhanced accordinly.

#### How to verify it
1. Verified that all pytest UT cases are passing.
2. Verified the route_check_test.sh script on single asic and multi-asic platforms.
3. Verified Monit routecheck outputs by simulating a failure scenario on both single asic and multi-asic platforms.

output from Monit Check:
**Single Asic:**
xxx/usr/local/bin# monit status routecheck
Monit 5.20.0 uptime: 1d 20h 32m

Program 'routeCheck'
  status                       Status failed
  monitoring status            Monitored
  monitoring mode              active
  on reboot                    start
  last exit value              255
  last output                  Failure results: {{
                                   "": {
                                       "missed_ROUTE_TABLE_routes": [
                                           "20c0:d9b8:99:80::/64"
                                       ]
                                   }
                               }}
                               Failed. Look at reported mismatches above
                               add: {
                                   "": []
                               }
                               del: {
                                   "": []
                               }
  data collected               Tue, 12 Dec 2023 20:30:11

'''
**Multi Asic:**
'''
/bin# monit status routecheck
Monit 5.20.0 uptime: 1d 23h 51m

Program 'routeCheck'
  status                       Status failed
  monitoring status            Monitored
  monitoring mode              active
  on reboot                    start
  last exit value              255
  last output                  Failure results: {{
                                   "asic0": {
                                       "missed_ROUTE_TABLE_routes": [
                                           "1.0.0.0/16"
                                       ]
                                   },
                                   "asic1": {
                                       "missed_ROUTE_TABLE_routes": [
                                           "1.0.0.0/16"
                                       ]
                                   },
                                   "asic2": {
                                       "missed_ROUTE_TABLE_routes": [
                                           "1.0.0.0/16"
                                       ]
                                   }
                               }}
                               Failed. Look at reported mismatches above
                               add: {
                                   "asic0": [],
                                   "asic1": [],
                                   "asic2": []
                               }
                               del: {
                                   "asic0": [],
                                   "asic1": [],
                                   "asic2": []
                               }
  data collected               Tue, 12 Dec 2023 23:54:23
'''

### CODE REVIEW

File: scripts/route_check_test.sh [hunk 1]
```diff
@@ -2,36 +2,85 @@
...
-sonic-db-cli APPL_DB del "NEIGH_TABLE:Vlan1000:fc02:1000::99" > /dev/null
-sonic-db-cli ASIC_DB del 'ASIC_STATE:SAI_OBJECT_TYPE_ROUTE_ENTRY:{"dest":"fc02:1000::99/128","switch_id":"oid:0x21000000000000","vr":"oid:0x300000000007c"}' > /dev/null
-
-echo "------"
-echo "expect success!"
-echo "Running Route Check..."
-./route_check.py
-echo "return value: $?"
+
+CONFIG_FILE="/etc/sonic/config_db.json"
```
Comment (arlakshm) [inline scripts/route_check_test.sh:6]: how is the test run? if config_db is not present will it awk command will fail?

**Review by arlakshm — COMMENTED** [review]
as comments

**Comment by deepak-singhal0408** [comment]
Thanks @arlakshm for the review comments..
JFYI, other than that, I have addressed the packet-chassis issue by skipping the eth1 routes in APPL_DB.. Also added a multi-asic testcase for frr routes mismatch..

Please have a look and let me know if anything.

**Review by abdosi — APPROVED** [review]
LGTM

File: scripts/route_check.py [hunk 2]
```diff
@@ -687,85 +692,102 @@ def check_routes():
...
     """
-    intf_appl_miss = []
-    rt_appl_miss = []
-    rt_asic_miss = []
-    rt_frr_miss = []
+    namespace_list = []
+    if namespace is not multi_asic.DEFAULT_NAMESPACE and namespace in multi_asic.get_namespace_list():
+        namespace_list.append(namespace)
+    else:
+        namespace_list = multi_asic.get_namespace_list()
```
Comment (judyjoseph) [inline scripts/route_check.py:699]: In MULTI-ASIC devices,   can you check if get_namespace_list() returns DEFAULT_NAMESPACE which is ''. if so we can skip that

File: scripts/route_check.py [hunk 3]
```diff
@@ -687,85 +692,102 @@ def check_routes():
...
+            results[namespace]["Unaccounted_ROUTE_ENTRY_TABLE_entries"] = rt_asic_miss
 
-    rt_frr_miss = check_frr_pending_routes()
+        rt_frr_miss = check_frr_pending_routes(namespace)
 
-    if rt_frr_miss:
-        results["missed_FRR_routes"] = rt_frr_miss
+        if rt_frr_miss:
+            if namespace not in results:
+                results[namespace] = {}
```
Comment (judyjoseph) [inline scripts/route_check.py:772]: With this namespace looping approach, are we altering the results array ? Just to make sure for single asic route_check.py still is backward compatible

**Review by judyjoseph — APPROVED** [review]

**Comment by StormLiangMS** [comment]
@deepak-singhal0408 could you run test with 202305 with this PR?

**Comment by deepak-singhal0408** [comment]
Hi @StormLiangMS , ran test on T1 DUT running 20230531.12.
Results are fine..

Results:
'''
ABZNC1 ESXJMW W4BHYA 73X1AJ ACAFHR BSS91C 8XF435 TTJGD9
DUT:/usr/local/bin# route_check.py -m INFO
Checking routes for namespaces: ['']
All good!
:/usr/local/bin# route_check.py
:/usr/local/bin# mv route_check_test.sh route_check_test.sh1
:/usr/local/bin# touch route_check_test.sh
:/usr/local/bin# chmod +x route_check_test.sh
:/usr/local/bin# vim route_check_test.sh
:/usr/local/bin# route_check_test.sh
Platform: xxxxxxxx
num_asic: 1
------
expect errors!
Running Route Check...
Failure results: {{
    "": {
        "missed_ROUTE_TABLE_routes": [
            "20c0:d9b8:99:80::/64"
        ],
        "missed_INTF_TABLE_entries": [
            "10.0.0.99/32"
        ],
        "Unaccounted_ROUTE_ENTRY_TABLE_entries": [
            "192.193.120.255/25"
        ]
    }
}}
Failed. Look at reported mismatches above
add: {
    "": []
}
del: {
    "": []
}
return value: 255
------
expect success on dualtor, expect error on all other devices!
Running Route Check...
Failure results: {{
    "": {
        "Unaccounted_ROUTE_ENTRY_TABLE_entries": [
            "fc02:1000::99/128"
        ]
    }
}}
Failed. Look at reported mismatches above
add: {
    "": []
}
del: {
    "": []
}
return value: 255
------
expect success!
Running Route Check...
return value: 0
'''
'''

**Comment by deepak-singhal0408** [comment]
Hi @StormLiangMS , ran the test on a T1 running 202305.12
Results are fine..

Results:

/usr/local/bin# route_check.py -m INFO
Checking routes for namespaces: ['']
All good!
usr/local/bin# route_check_test.sh
Platform: xxxxxxxx
num_asic: 1
expect errors!
Running Route Check...
Failure results: {{
    "": {
        "missed_ROUTE_TABLE_routes": [
            "20c0:d9b8:99:80::/64"
        ],
        "missed_INTF_TABLE_entries": [
            "10.0.0.99/32"
        ],
        "Unaccounted_ROUTE_ENTRY_TABLE_entries": [
            "192.193.120.255/25"
        ]
    }
}}
Failed. Look at reported mismatches above
add: {
    "": []
}
del: {
    "": []
}
return value: 255

expect success on dualtor, expect error on all other devices!
Running Route Check...
Failure results: {{
    "": {
        "Unaccounted_ROUTE_ENTRY_TABLE_entries": [
            "fc02:1000::99/128"
        ]
    }
}}
Failed. Look at reported mismatches above
add: {
    "": []
}
del: {
    "": []
}
return value: 255

expect success!
Running Route Check...
return value: 0

**Comment by StormLiangMS** [comment]
hi @deepak-singhal0408 cherry pick conflict, could you file separate PR for 202305?

**Comment by yxieca** [comment]
@anamehra @deepak-singhal0408 https://github.com/sonic-net/sonic-utilities/pull/3075 and https://github.com/sonic-net/sonic-utilities/pull/3077 have blocked sonic-utilities repo sub-module advance in sonic-buildimage repo.

**Pull request merged**